src/jtag/drivers/remote_bitbang.c

```diff
@@ -51,10 +51,48 @@ static bool remote_bitbang_buf_full(void)
 		 sizeof(remote_bitbang_recv_buf));
 }
 
+static int remote_bitbang_flush(void)
+{
+	if (remote_bitbang_send_buf_used <= 0)
+		return ERROR_OK;
+
+	unsigned int offset = 0;
+	while (offset < remote_bitbang_send_buf_used) {
+		ssize_t written = write_socket(remote_bitbang_fd, remote_bitbang_send_buf + offset,
+									   remote_bitbang_send_buf_used - offset);
+		if (written < 0) {
+			log_socket_error("remote_bitbang_putc");
+			remote_bitbang_send_buf_used = 0;
+			return ERROR_FAIL;
+		}
+		offset += written;
+	}
+	remote_bitbang_send_buf_used = 0;
+	return ERROR_OK;
+}
+
+typedef enum {
+	NO_BLOCK,
+	BLOCK
+} block_bool_t;
+
 /* Read any incoming data, placing it into the buffer. */
-static int remote_bitbang_fill_buf(void)
+static int remote_bitbang_fill_buf(block_bool_t block)
 {
-	socket_nonblock(remote_bitbang_fd);
+	if (remote_bitbang_recv_buf_start == remote_bitbang_recv_buf_end) {
+		/* If the buffer is empty, reset it to 0 so we get more
+		 * contiguous space. */
+		remote_bitbang_recv_buf_start = 0;
+		remote_bitbang_recv_buf_end = 0;
+	}
+
+	if (block == BLOCK) {
+		if (remote_bitbang_flush() != ERROR_OK)
+			return ERROR_FAIL;
+		socket_block(remote_bitbang_fd);
+	}
+
+	bool first = true;
 	while (!remote_bitbang_buf_full()) {
 		unsigned int contiguous_available_space;
```

Comment:
The computation of contiguous_available_space could also go to a separate function (for code readability).